--- compiler/generator/interpreter/fbc_interpreter.hh	2020-12-23 22:09:06.000000000 +0100
+++ compiler/generator/interpreter/fbc_interpreter.hh	2021-07-27 12:10:28.014913841 +0200
@@ -612,7 +612,9 @@
 #define emptyReturnScal() (addr_stack_index == 0)
 
         // Check block coherency
-        block->check();
+        if (TRACE > 0) {
+          block->check();
+        }
 
         InstructionIT it = block->fInstructions.begin();
         dispatchFirstScal();
