--- compiler/generator/llvm/llvm_dynamic_dsp_aux.cpp~	2019-12-22 18:23:58.583809298 +0100
+++ compiler/generator/llvm/llvm_dynamic_dsp_aux.cpp	2020-01-08 15:38:53.420330443 +0100
@@ -413,7 +413,8 @@
             argv1[argc1++] = "string";
             // Filter arguments
             for (int i = 0; i < argc; i++) {
-                if (!testArg(argv[i])) argv1[argc1++] = argv[i];
+              argv1[argc1++] = argv[i];
+              //                if (!testArg(argv[i])) argv1[argc1++] = argv[i];
             }
             argv1[argc1] = nullptr;  // NULL terminated argv
             
