diff -cr libxcb-1.12-old/src/c_client.py libxcb-1.12/src/c_client.py
*** libxcb-1.12-old/src/c_client.py	2017-06-04 11:33:44.502091374 +0200
--- libxcb-1.12/src/c_client.py	2017-06-04 11:35:40.050900543 +0200
***************
*** 1934,1943 ****
  		    # compute the length of it by subtracting it from
                      # the adress of the first byte after the end of the
                      # request
! 		    after_end_of_request = '(((char*)R) + R->length * 4)'
! 		    start_of_list = '%s(R)' % (field.c_accessor_name)
                      bytesize_of_list = '%s - (char*)(%s)' % (after_end_of_request, start_of_list)
! 		    return '(%s) / sizeof(%s)' % (bytesize_of_list, field.type.member.c_wiretype)
              else:
                  raise Exception(
                      "lengthless lists with varsized members are not supported. Fieldname '%s'"
--- 1934,1943 ----
  		    # compute the length of it by subtracting it from
                      # the adress of the first byte after the end of the
                      # request
!                     after_end_of_request = '(((char*)R) + R->length * 4)'
!                     start_of_list = '%s(R)' % (field.c_accessor_name)
                      bytesize_of_list = '%s - (char*)(%s)' % (after_end_of_request, start_of_list)
!                     return '(%s) / sizeof(%s)' % (bytesize_of_list, field.type.member.c_wiretype)
              else:
                  raise Exception(
                      "lengthless lists with varsized members are not supported. Fieldname '%s'"
