--- qsciscintillabase.cpp~	2016-04-18 11:08:48.000000000 +0200
+++ qsciscintillabase.cpp	2016-06-22 11:57:22.353636841 +0200
@@ -89,6 +89,7 @@
         , clickCausedFocus(false)
 #endif
 {
+
     connect(verticalScrollBar(), SIGNAL(valueChanged(int)),
             SLOT(handleVSb(int)));
 
@@ -134,6 +135,10 @@
 
     // Add it to the pool.
     poolList.append(this);
+
+    setLineWidth(0);
+    setMidLineWidth(0);
+    setFrameShape(QFrame::NoFrame);
 }
 
 
