diff --git a/pluginhost/JuceLibraryCode/modules/juce_core/native/juce_ComSmartPtr_windows.h b/pluginhost/JuceLibraryCode/modules/juce_core/native/juce_ComSmartPtr_windows.h
index 0a24a7c23..32b30e79e 100644
--- a/pluginhost/JuceLibraryCode/modules/juce_core/native/juce_ComSmartPtr_windows.h
+++ b/pluginhost/JuceLibraryCode/modules/juce_core/native/juce_ComSmartPtr_windows.h
@@ -23,7 +23,7 @@
 namespace juce
 {
 
-#if (JUCE_MINGW && JUCE_32BIT) || (! defined (_MSC_VER) && ! defined (__uuidof))
+#if 0 //(JUCE_MINGW && JUCE_32BIT) || (! defined (_MSC_VER) && ! defined (__uuidof))
  #ifdef __uuidof
   #undef __uuidof
  #endif
