diff -ruN old/pure-data/tcl/pd_menus.tcl new/pure-data/tcl/pd_menus.tcl
--- old/pure-data/tcl/pd_menus.tcl	2026-09-28 21:33:31.526186932 +0200
+++ new/pure-data/tcl/pd_menus.tcl	2026-09-28 21:33:31.526186932 +0200
@@ -236,6 +236,15 @@
     $mymenu add command -label [_ "Comment"]  -accelerator "$accelerator+5" \
         -command {::pd_menucommands::scheduleAction menu_send_float $::focused_window text 0}
     $mymenu add  separator
+
+    $mymenu add command -label [_ "Radium Float"] \
+        -command {::pd_menucommands::scheduleAction menu_send_float $::focused_window obj "-16000 -16000 r_float <set name here>"}
+    $mymenu add command -label [_ "Radium Integer"] \
+        -command {::pd_menucommands::scheduleAction menu_send_float $::focused_window obj "-16000 -16000 r_int <set name here>"}
+    $mymenu add command -label [_ "Radium Toggle"] \
+        -command {::pd_menucommands::scheduleAction menu_send_float $::focused_window obj "-16000 -16000 r_bool <set name here>"}
+    $mymenu add  separator
+
     $mymenu add command -label [_ "Bang"]     -accelerator "Shift+$accelerator+B" \
         -command {::pd_menucommands::scheduleAction menu_send $::focused_window bng}
     $mymenu add command -label [_ "Toggle"]   -accelerator "Shift+$accelerator+T" \
@@ -259,6 +268,46 @@
         -command {::pd_menucommands::scheduleAction menu_send $::focused_window graph}
     $mymenu add command -label [_ "Array"]    -accelerator "Shift+$accelerator+A" \
         -command {::pd_menucommands::scheduleAction menu_send $::focused_window menuarray}
+    $mymenu add  separator
+
+    $mymenu add command -label [_ "Radium Note On"] \
+        -command {::pd_menucommands::scheduleAction menu_send_float $::focused_window obj "-16000 -16000 r_note_on"}
+    $mymenu add command -label [_ "Radium Note Off"] \
+        -command {::pd_menucommands::scheduleAction menu_send_float $::focused_window obj "-16000 -16000 r_note_off"}
+    $mymenu add command -label [_ "Radium Notes"] \
+        -command {::pd_menucommands::scheduleAction menu_send_float $::focused_window obj "-16000 -16000 r_notes"}
+    $mymenu add command -label [_ "Radium Note Pitch"] \
+        -command {::pd_menucommands::scheduleAction menu_send_float $::focused_window obj "-16000 -16000 r_pitch"}
+    $mymenu add command -label [_ "Radium Note Velocity"] \
+        -command {::pd_menucommands::scheduleAction menu_send_float $::focused_window obj "-16000 -16000 r_velocity"}
+    $mymenu add command -label [_ "Radium Send Note On"] \
+        -command {::pd_menucommands::scheduleAction menu_send_float $::focused_window obj "-16000 -16000 r_send_note_on"}
+    $mymenu add  separator
+
+    $mymenu add command -label [_ "Radium Send Note Off"] \
+        -command {::pd_menucommands::scheduleAction menu_send_float $::focused_window obj "-16000 -16000 r_send_note_off"}
+    $mymenu add command -label [_ "Radium Send Note Pitch"] \
+        -command {::pd_menucommands::scheduleAction menu_send_float $::focused_window obj "-16000 -16000 r_send_pitch"}
+    $mymenu add command -label [_ "Radium Send Note Velocity"] \
+        -command {::pd_menucommands::scheduleAction menu_send_float $::focused_window obj "-16000 -16000 r_send_velocity"}
+    $mymenu add  separator
+
+    $mymenu add command -label [_ "Radium Pipe Note On"] \
+        -command {::pd_menucommands::scheduleAction menu_send_float $::focused_window obj "-16000 -16000 r_pipe_note_ons"}
+    $mymenu add command -label [_ "Radium Pipe Note Off"] \
+        -command {::pd_menucommands::scheduleAction menu_send_float $::focused_window obj "-16000 -16000 r_pipe_note_offs"}
+    $mymenu add command -label [_ "Radium Pipe Note Velocity"] \
+        -command {::pd_menucommands::scheduleAction menu_send_float $::focused_window obj "-16000 -16000 r_pipe_velocity"}
+    $mymenu add command -label [_ "Radium Pipe Note Pitch"] \
+        -command {::pd_menucommands::scheduleAction menu_send_float $::focused_window obj "-16000 -16000 r_pipe_pitch"}
+    $mymenu add  separator
+
+    $mymenu add command -label [_ "Radium Line"] \
+        -command {::pd_menucommands::scheduleAction menu_send_float $::focused_window obj "-16000 -16000 r_line"}
+    $mymenu add command -label [_ "Radium Visible Line"] \
+        -command {::pd_menucommands::scheduleAction menu_send_float $::focused_window obj "-16000 -16000 r_visibleline"}
+    $mymenu add command -label [_ "Radium Set Block Reltempo"] \
+        -command {::pd_menucommands::scheduleAction menu_send_float $::focused_window obj "-16000 -16000 r_send_blockreltempo"}
 }
 
 proc ::pd_menus::build_find_menu {mymenu {patchwindow true}} {
